File: app/Day5.hs
import Debug.Trace

main :: IO ()
main = do
    contents <- readFile "inputs/day5.txt"
    let listOfWords = map words $ lines contents
    let tpls = endStart listOfWords
    --print $ show tpls
    let horz = sortHorz tpls 
    --print horz
    let verts = sortVert tpls 
    --print verts
    print "Part 1"
    let merged = merge horz verts
    --print merged
    let coordsHit = concatMap generateCoords merged
    --print coordsHit
    print $ filter (\x -> 1 < snd x) $ getOccurences (uniq coordsHit) coordsHit
    print $ length $ filter (\x -> 1 < snd x) $ getOccurences (uniq coordsHit) coordsHit
    -- part due

merge :: [a] -> [a] -> [a]
merge [] ys = ys
merge (x:xs) ys = x:merge ys xs

endStart :: [[String]] -> [((Int,Int), (Int,Int))]
endStart [] = []
endStart (tup:xs) = ((read  (head x), read  (last x)),(read  (head y), read  (last y))):endStart xs
    where   x = words $ repl $ head tup :: [String]
            y = words $ repl $ last tup :: [String]
            repl = replace (==',') ' '


replace :: (Char -> Bool) -> Char -> String -> String
replace _ _ [] = []
replace con rep (c:cs) = if con c then rep:replace con rep cs else c:replace con rep cs


sortHorz ::  [((Int,Int), (Int,Int))] ->  [((Int,Int), (Int,Int))]
sortHorz [] = []
sortHorz (coords@((a,_),(b,_)):xs) 
    | a == b  = coords : sortHorz xs 
    | otherwise= sortHorz xs 

sortVert :: [((Int,Int), (Int,Int))] ->  [((Int,Int), (Int,Int))]
sortVert [] = []
sortVert (coords@((_,a),(_,b)):xs) 
    | a == b = coords : sortVert xs
    | otherwise = sortVert xs

generateCoords :: ((Int,Int),(Int, Int)) -> [(Int,Int)]
generateCoords ((a1,a2),(b1,b2)) = zipB a b
    where   a = if a1 < b1 then [a1 .. b1] else reverse [b1 .. a1]
            b = if a2 < b2 then [a2 .. b2] else reverse [b2 .. a2]

zipB [a] ls = [(a,l) | l <- ls]
zipB ls [a] = [(l,a) | l <- ls]
zipB hs ls = zip hs ls

getOccurences [] _  = []
getOccurences (u:us) xs = (u, length $ filter (== u) xs):getOccurences us xs

uniq :: Eq a => [a] -> [a]
uniq [] = []
uniq (x:xs) = (if x `elem` xs then id else (x:)) $ uniq xs

File: app/Day5.py
# %%
strings = [line.strip() for line in open('../inputs/day5.txt')]

# %%
def tuplefy(coord):
    a,b = x if len(x := coord.split(',')) == 2 else print(x)
    return (int(a), int(b))
linecoords = [[tuplefy(coords)  for coords in line.split(" -> ")] for line in strings]
# %%
def straight(coords):
    (a,b),(c,d)=coords
    return (a == c) or (b==d)
straights = [coords for coords in linecoords if straight(coords)]
# %%
def generateLine(coords):
    a,b = coords
    xs = list(range(a[0],b[0]+1)) if a[0] < b[0] else list(range(b[0],a[0]+1))[::-1]
    ys = list(range(a[1],b[1]+1)) if a[1] < b[1] else list(range(b[1],a[1]+1))[::-1]
    line = []
    if len(xs)==1:
        line = [(xs[0],y) for y in ys]
    elif len(ys)==1:
        line = [(x, ys[0]) for x in xs]
    else:
        line = zip(xs,ys)
    return line


# %%
def countHits(lines):
    coordsHit =  {}
    coordWithMoreThanToo = 0
    for st in lines:
        line = generateLine(st)
        for point in line:
            if point in coordsHit:
                if coordsHit[point] == 1: coordWithMoreThanToo += 1
                coordsHit[point] = coordsHit[point] + 1
            else:
                coordsHit[point] = 1
    return coordWithMoreThanToo
print(countHits(straights))
# %%
print(countHits(linecoords))